app/src/main/java/org/dharmaseed/android/TalkRepository.java

```diff
@@ -185,6 +187,36 @@ public Cursor getTalksByCenter(
                 DBManager.C.Center.TABLE_NAME + "." + DBManager.C.Center.ID + "=" + venueId);
     }
 
+    /** Check if we have talks downloaded in old/inaccessible locations, and mark them as not downloaded if so
+     *
+     */
+    public void removeOldDownloads()
+    {
+        File downloadsDir = TalkManager.getDir(dbManager.getContext());
+        Log.i(LOG_TAG, "HI " + downloadsDir.toString());
+
+        ArrayList<String> columns = new ArrayList<String>();
+        columns.add(DBManager.C.Talk.TABLE_NAME + "." + DBManager.C.Talk.ID);
+        columns.add(DBManager.C.Talk.TABLE_NAME + "." + DBManager.C.Talk.FILE_PATH);
+
+        Cursor downloaded = getTalks(columns, null, false, true);
+
+        while (downloaded.moveToNext())
+        {
+            int id = downloaded.getInt(downloaded.getColumnIndexOrThrow(DBManager.getAlias(
+                    DBManager.C.Talk.TABLE_NAME + "." + DBManager.C.Talk.ID)));
+            String path = downloaded.getString(downloaded.getColumnIndexOrThrow(DBManager.getAlias(
+                    DBManager.C.Talk.TABLE_NAME + "." + DBManager.C.Talk.FILE_PATH))).trim();
+
+            String fullPath = new File(path).getAbsolutePath();
+            String downloadPath = TalkManager.getDir(dbManager.getContext()).getAbsolutePath();
+
+            if (! fullPath.startsWith(downloadPath)) {
+                Log.w(LOG_TAG, "Removing old download for talk " + id);
+                dbManager.removeDownload(id);
```

Comment:
this removes our log for it in the DB, but it leaves the file on their system doesn't it? Can we also delete the file from the system? Or, even better, move the file from the old dir to the new one?

I would hate for people to suddenly lose all of their downloaded talks with a new update, although it seems like they haven't been able to play them for a little while anyway. What are your thoughts?

Reply:
Yes, that's a great idea, thanks!  I've updated it to attempt to move the file to the new location, and if that succeeds, to re-add it to the downloads table.  I've tested this to verify that it does indeed preserve previously downloaded files when upgrading the app.